atc/exec/set_pipeline_step.go

```diff
@@ -0,0 +1,244 @@
+
+	Initializing(lager.Logger)
+	Starting(lager.Logger)
+	Finished(lager.Logger, ExitStatus, db.ConfigVersion, db.ConfigVersion)
+
+	FindTeam() (db.Team, bool, error)
+}
+
+// SetPipelineStep sets a pipeline to current team. If pipeline_name specified
+// is "self", then it will self set the current pipeline. This step takes pipeline
+// configure file and var files from some resource in the pipeline, like git.
+type SetPipelineStep struct {
+	planID      atc.PlanID
+	plan        atc.SetPipelinePlan
+	metadata    StepMetadata
+	delegate    SetPipelineDelegate
+	lockFactory lock.LockFactory
+	succeeded   bool
+}
+
+func NewSetPipelineStep(
+	planID atc.PlanID,
+	plan atc.SetPipelinePlan,
+	metadata StepMetadata,
+	delegate SetPipelineDelegate,
+	lockFactory lock.LockFactory,
+) Step {
+	return &SetPipelineStep{
+		planID:   planID,
+		plan:     plan,
+		metadata: metadata,
+		delegate:    delegate,
+		lockFactory: lockFactory,
+	}
+}
+
+func (step *SetPipelineStep) Run(ctx context.Context, state RunState) error {
+	logger := lagerctx.FromContext(ctx)
+	logger = logger.Session("set-pipeline-step", lager.Data{
+		"step-name": step.plan.Name,
+		"job-id":    step.metadata.JobID,
+	})
+
+	step.delegate.Initializing(logger)
+
+	variables := step.delegate.Variables()
+	params, err := creds.NewParams(variables, step.plan.Params).Evaluate()
+	if err != nil {
+		return err
+	}
+
+	spParams, err := atc.NewSetPipelineParams(params)
+	if err != nil {
+		return err
+	}
+
+	repository := state.Artifacts()
+	artifactSource := setPipelineArtifactSource{
+		ctx:    ctx,
+		logger: logger,
+		params: spParams,
+		repo:   repository,
+	}
+
+	atcConfig, err := artifactSource.FetchConfig()
+	if err != nil {
+		return err
+	}
+
+	step.delegate.Starting(logger)
+
+	stdout := step.delegate.Stdout()
+
+	if spParams.PipelineName == "self" {
+		spParams.PipelineName = step.metadata.PipelineName
+	}
+
+	team, found, err := step.delegate.FindTeam()
+	if err != nil {
+		logger.Error("failed-to-find-team", err)
+		return err
+	}
+	if !found {
+		err = fmt.Errorf("team not found")
+		logger.Error("failed-to-find-team", err)
+		return err
+	}
+
+	fromVersion := db.ConfigVersion(0)
+	pipeline, found, err := team.Pipeline(spParams.PipelineName)
+	if err != nil {
+		return nil
+	}
+
+	var existingConfig atc.Config
+	if !found {
+		existingConfig = atc.Config{}
+	} else {
+		fromVersion = pipeline.ConfigVersion()
+		existingConfig, err = pipeline.Config()
+		if err != nil {
+			return err
+		}
+	}
+
+	diffExists := existingConfig.Diff(stdout, *atcConfig)
+	if !diffExists {
+		logger.Debug("no-diff")
+
+		if spParams.FailWhenNoDiff {
+			return errors.New("no diff found")
+		}
+
+		fmt.Fprintf(stdout, "No diff found.\n")
+		step.succeeded = true
+		step.delegate.Finished(logger, ExitStatus(0), fromVersion, fromVersion)
+		return nil
+	}
+
+	fmt.Fprintf(stdout, "Updating the pipeline.\n")
+	pipeline, _, err = team.SavePipeline(spParams.PipelineName, *atcConfig, fromVersion, false)
```

Comment:
Per the original proposal, this should be using the step name as the pipeline name: https://github.com/vito/rfcs/blob/set-pipeline-step/031-set-pipeline-step/proposal.md#proposal

e.g.:

```yaml
set_pipeline: foo
file: ci/pipelines/foo.yml
```

In v10, assuming you have a project called `ci`, this could be shortened further to:

```yaml
set_pipeline: foo
```

With your proposed `self` behavior, it would just be `set_pipeline: self`.

How does that sound? :slightly_smiling_face:

Reply:
Sounds good.